src/util/find_symbols.cpp

```diff
@@ -201,103 +243,148 @@ static bool find_symbols(
-  });
+  find_symbols_in_expr_and_subs(
+    symbol_kindt::F_TYPE,
+    src,
+    [&dest](const symbol_exprt &e) {
+      dest.insert(e.get_identifier());
+      return true;
+    },
+    {});
 }
 
 void find_type_symbols(const typet &src, find_symbols_sett &dest)
 {
-  find_symbols(symbol_kindt::F_TYPE, src, [&dest](const symbol_exprt &e) {
-    dest.insert(e.get_identifier());
-    return true;
-  });
+  find_symbols_in_type_and_subs(
+    symbol_kindt::F_TYPE,
+    src,
+    [&dest](const symbol_exprt &e) {
+      dest.insert(e.get_identifier());
+      return true;
+    },
+    {});
 }
 
-void find_non_pointer_type_symbols(
-  const exprt &src,
-  find_symbols_sett &dest)
+void find_non_pointer_type_symbols(const exprt &src, find_symbols_sett &dest)
 {
-  find_symbols(
-    symbol_kindt::F_TYPE_NON_PTR, src, [&dest](const symbol_exprt &e) {
+  find_symbols_in_expr_and_subs(
+    symbol_kindt::F_TYPE_NON_PTR,
+    src,
+    [&dest](const symbol_exprt &e) {
       dest.insert(e.get_identifier());
       return true;
-    });
+    },
+    {});
 }
 
-void find_non_pointer_type_symbols(
-  const typet &src,
-  find_symbols_sett &dest)
+void find_non_pointer_type_symbols(const typet &src, find_symbols_sett &dest)
 {
-  find_symbols(
-    symbol_kindt::F_TYPE_NON_PTR, src, [&dest](const symbol_exprt &e) {
+  find_symbols_in_type_and_subs(
+    symbol_kindt::F_TYPE_NON_PTR,
+    src,
+    [&dest](const symbol_exprt &e) {
       dest.insert(e.get_identifier());
       return true;
-    });
+    },
+    {});
+}
+
+void find_type_and_expr_symbols(
+  const exprt &src,
+  find_symbols_sett &dest,
+  const std::unordered_set<irep_idt> &subs_to_find)
+{
+  find_symbols_in_expr_and_subs(
+    symbol_kindt::F_ALL,
+    src,
+    [&dest](const symbol_exprt &e) {
+      dest.insert(e.get_identifier());
+      return true;
+    },
+    subs_to_find);
 }
 
 void find_type_and_expr_symbols(const exprt &src, find_symbols_sett &dest)
 {
-  find_symbols(symbol_kindt::F_ALL, src, [&dest](const symbol_exprt &e) {
-    dest.insert(e.get_identifier());
-    return true;
-  });
+  find_type_and_expr_symbols(src, dest, {});
+}
+
+void find_type_and_expr_symbols(
+  const typet &src,
+  find_symbols_sett &dest,
+  const std::unordered_set<irep_idt> &subs_to_find)
+{
+  find_symbols_in_type_and_subs(
+    symbol_kindt::F_ALL,
+    src,
+    [&dest](const symbol_exprt &e) {
+      dest.insert(e.get_identifier());
+      return true;
+    },
+    subs_to_find);
 }
 
 void find_type_and_expr_symbols(const typet &src, find_symbols_sett &dest)
 {
-  find_symbols(symbol_kindt::F_ALL, src, [&dest](const symbol_exprt &e) {
-    dest.insert(e.get_identifier());
-    return true;
-  });
+  find_type_and_expr_symbols(src, dest, {});
 }
 
 void find_symbols(const exprt &src, find_symbols_sett &dest)
 {
-  find_symbols(symbol_kindt::F_EXPR, src, [&dest](const symbol_exprt &e) {
-    dest.insert(e.get_identifier());
-    return true;
-  });
+  find_symbols_in_expr_and_subs(
```

Comment:
You could substantially reduce the churn by adding a variant of `find_symbols_in_expr_and_subs` that either doesn't require the "subs" argument, or uses a default value of `{}`. By extension of this I'd argue that you can reduce churn even further by just making `find_symbols` have this capability.

Reply:
Sounds great. I made the `static find_symbols` take the "subs" argument without default value and the public function `find_type_and_expr_symbols` take the `subs` argument with the default value of `{}`.

Reply:
Hmm, I think if you did it the other way around (make the non-public function use a default value) you could substantially reduce the size of the diff in this PR. Also, I don't think the default value on the (new!) public function adds as much value for any use of that function without the third argument just amounts to using one of the other, pre-existing functions.

Reply:
I moved the default value to the static function and it actually reduce the diff of this PR. 
The public function with the new `subs_to_find` argument is different from other pre-existing functions. It calls the static function `find_symbols` with kind `symbol_kindt::F_ALL` while others call `find_symbols` with different kind arguments.